Add support for installing content with dnf

gtkramer opened the pull request:
In test scenarios, work to create a swupd update stream is not needed
and increases image generation time.  If content for a test scenario can
be supplied as a package repository, then dnf can be used to do an
install of packages in the same way mixer generates chroots.

To support a define-and-run approach with ister, template files may
define an additional "PackageManager" key, which may be either "dnf" or
"swupd", if not defined to preserve backwards compatbility.  In the case
where the package manager is swupd, the list of bundles remains the same
as it is today.  In the case where the package manager is dnf, the
bundle list is treated as a package list that is used during a dnf
install.

Much like how content and version URLs may be passed at the command line
for swupd, to pass package repository URLs, which there may be multiple,
a DNF configuration file must be passed instead with the new
--dnf-config option.

To mimic how swupd works after installing content, update triggers are
run if they exist.  Because using DNF makes ister distro-agnostic,
opting to make the image bootable by using a post non-chroot script as
opposed to calling clr-boot-manager update directly.

Signed-off-by: George T Kramer <george.t.kramer@intel.com>

gtkramer commented:
Unit tests still need updated.  Looking to get some early feedback on the initial approach.

bryteise commented on dnf-boot.sh:
```diff
@@ -0,0 +1,2 @@
+#!/bin/bash
+"$1"/usr/bin/clr-boot-manager update -p "$1"
```
Inclined to call the file dnf-post-cbm.sh

gtkramer replied:
Maybe pkg-post-update.sh?  Since it would also be running the update helper and cbm update?  pkg since in theory, we could add support for additional package managers?

bryteise replied:
Ah yes sure

bryteise commented on ister.py:
```diff
@@ -408,43 +411,62 @@ def get_current_format():
+        cmd += " --statedir={0}".format(args.statedir)
+        if args.cert_file:
+            cmd += " -C {0}".format(args.cert_file)
+    else:
```
I'm inclined to match specificly and error.

gtkramer replied:
Match specifically?  Maybe we could default the cert file to the one on the system for swupd?

bryteise replied:
Ah sorry I mean if SWUPD elif DNF else raise error

gtkramer replied:
Ah, yessir!  I'll add this in.

bryteise commented on ister.py:
```diff
@@ -408,43 +411,62 @@ def get_current_format():
         frmt = format_file.read().strip()
     return frmt
 
+def add_default_dnf_config(args):
```
I'm inclined to not have this built into ister and just let dnf use /etc/dnf.conf if it exists by default or else error early in the process around validate template.

bryteise replied:
Or if the dnf-conf passed in doesn't exist still error.

gtkramer replied:
Having this built-in would allow for a similar default for dnf analogous to what swupd has with its content and version URLs.  It's kinda ugly though, but maybe less ugly than the alternative of passing repo URLs and then generating a conf in /tmp somewhere?

bryteise replied:
So the built-ins for swupd urls are essentially legacy.
It would be nice to remove them and None the argument by default. Then we can only add those arguments to swupd when they are provided instead of every time.

gtkramer replied:
Ok.  I'll do a separate commit for this as pre-work for this commit.

bryteise commented on ister.py:
```diff
@@ -408,43 +411,62 @@ def get_current_format():
+    LOG.info("Starting {0}. May take several minutes".format(package_manager))
+
+    if package_manager == SWUPD:
+        if not args.format:
```
Probably would reorganize this section in a different commit.

gtkramer replied:
Maybe I could break this out into two different functions?  Something like
```
def install_os(args):
    if package_manager == SWUPD:
        install_os_swupd(args)
    elif package_manager == DNF:
        install_os_dnf(args)
```

bryteise replied:
That'd be fine but I meant the reorganization of the swupd options being appended to the cmd sorry.

gtkramer replied:
I'll create a few different commits to organize things a little more nicely as a result of supporting multiple package managers.

bryteise commented on ister.py:
```diff
@@ -408,43 +411,62 @@ def get_current_format():
-        run_command("mount --bind {0}/var/tmp {1}"
-                    .format(target_dir, args.statedir))
-    swupd_env = os.environ
+        run_command("mount --bind {0}/var/tmp {1}".format(target_dir, args.statedir))
```
Why reformat this line?

gtkramer replied:
No need at all :)

bryteise commented on ister.py:
```diff
@@ -408,43 +411,62 @@ def get_current_format():
+    if package_manager == SWUPD and args.fast_install:
         run_command("rm -rf {0}".format(args.statedir))
+    if package_manager == DNF and os.path.exists("{0}/usr/bin/update-helper".format(target_dir)):
+        run_command("{0}/usr/bin/update-helper {1}".format(target_dir, target_dir))
```
This should be part of the post-dnf script.

gtkramer replied:
Ah, yes.  I'll move this in.

bryteise commented on ister.py:
```diff
@@ -768,7 +790,12 @@ def get_template(template_location):
     json_file = request.urlopen(template_location)
-    return json.loads(json_file.read().decode("utf-8"))
+    parsed_json = json.loads(json_file.read().decode("utf-8"))
+    try:
```
I think better to just do ->
if not parsed_json.get("PackageManager"):
    parsed_json["PackageManager"] = SWUPD

bryteise replied:
And I'm curious why this is done here instead of in validate_template?

gtkramer replied:
I was thinking of defaulting the PackageManager to swupd as close to the source as possible so that validate_template would be strictly checking if the template that was supplied was all right instead of possibly modifying it with a default value.

gtkramer replied:
Ah, so that's what the get function is used for.  I was initially doing `if not parsed_json["PackageManager"]` but getting a KeyError exception.  I Googled badly for that one.

bryteise replied:
Okay, that's fine with me then.

gtkramer commented:
I've made some additional commits that hopefully help to tidy up the code base.  I'll take a look at the failing unit tests tomorrow.

gtkramer commented:
Something else to maybe consider?  I noticed that as a result of doing a dnf install of the packages listed in bundle definitions alone, the resulting image is not quite equivalent to an image that a swupd verify --install creates.  Namely, the /usr/share/clear folder is missing.  Besides running `update-helper` and `clr-boot-manager update` after a dnf install to get a bootable image, is there anything else that *should* be done?

bryteise commented:
It will not be equivalent you are correct. If you want the bundles files in /usr/share/clear you should poke them in the post script but they aren't important for the dnf installation, just for using swupd.

gtkramer commented:
Before names are cemented in stone for all time, I was wondering if PackageManager is the best term for both dnf and swupd?  dnf is a package manager, but swupd is a bundle manager, right?  Maybe SoftwareManager would be more accurate?

bryteise commented:
SoftwareManager is fine

gtkramer commented:
All tests have been updated.  I don't think there's anything else on my end.  How do things look?

bryteise commented on ister.py:
```diff
@@ -413,8 +413,6 @@ def copy_os(args, template, target_dir):
     """Wrapper for running install command
     """
     LOG.info("Starting swupd. May take several minutes")
-    if not args.format:
```
Similarly to content and version urls, I think we can just remove the grabbing the format from the filesystem entirely.

gtkramer replied:
Ah, ok.  I'll take this out in the "Remove legacy definitions" commit

bryteise commented on ister.py:
```diff
@@ -414,20 +414,11 @@ def copy_os(args, template, target_dir):
-        swupd_command += " --versionurl={0}".format(args.versionurl)
-    swupd_command += " --format={0}".format(args.format)
+
+    add_bundles(template, target_dir)
```
This is happening twice now? Seems like a mistake.

gtkramer replied:
Only once, just for swupd

bryteise commented on ister.py:
```diff
@@ -431,12 +439,12 @@ def copy_os(args, template, target_dir):
         cmd += " --contenturl={0}".format(args.contenturl)
     if args.versionurl:
         cmd += " --versionurl={0}".format(args.versionurl)
+    cmd += " --format={0}".format(args.format)
```
Seems odd to move these two lines. Any reason?

gtkramer replied:
I split out my changes a little poorly across the commit, but the order was important to get the tests to pass because the command string changes otherwise

bryteise commented on ister.py:
```diff
@@ -447,6 +455,21 @@ def copy_os(args, template, target_dir):
         run_command("rm -rf {0}".format(args.statedir))
 
+
+def copy_os_dnf(args, template, target_dir):
```
Needs unit test

gtkramer replied:
I just added tests for the dnf command generation.

bryteise commented on ister.py:
```diff
@@ -412,9 +412,17 @@ def get_current_format():
+        copy_os_dnf(args, template, target_dir)
+
 
+def copy_os_swupd(args, template, target_dir):
```
Needs unit test.

gtkramer replied:
Could the existing unit tests be used here because we're defaulting to swupd? Or, maybe I should add a unit test for checking the new copy_os function works to choose the correct copy_os_xxx as expected?

bryteise replied:
The latter, we want to ensure the right one is picked, can stub out the actual call to copy_os_xxx in the test.

bryteise commented on ister.py:
```diff
@@ -412,9 +412,17 @@ def get_current_format():
     """
-    LOG.info("Starting swupd. May take several minutes")
-    add_bundles(template, target_dir)
+    package_manager = template["SoftwareManager"]
```
Should update unit test to call out specificly running swupd and add test for dnf.

gtkramer commented on ister.py:
```diff
@@ -962,6 +989,14 @@ def validate_version_template(template):
         raise Exception("Invalid version string (must be 'latest')")
 
 
+def validate_softmgr_template(template):
```
Ah, thank you.  I'll add a test for this.

gtkramer commented:
I've gone through and cleaned up all of the commits.  I think they now more logically group changes while minimizing changes done.  I've also added more tests to help with closing the testing gap.  Things look ready on my end.  Can you take another look?

bryteise commented on ister_test.py:
```diff
@@ -1321,20 +1321,6 @@ def add_bundles_good():
 
 
-@open_wrapper("good", "1\n")
-def get_current_format_good():
```
I expect the test needs to be removed from the tests array in this commit too.

gtkramer replied:
It's already taken out ;)

bryteise replied:
Ah, okay sorry I didn't see that X(.

bryteise commented on ister_test.py:
```diff
@@ -3126,7 +3126,23 @@ def validate_template_bad_version():
+                "PartitionMountPoints": [],
+                "Version": "20110"}
+    try:
+        ister.validate_template(template)
```
I expect this and the num test to call the validate_version_template function directly.

gtkramer replied:
Oops, missed this.  I'll make the change.

bryteise commented on ister.py:
```diff
@@ -428,16 +416,36 @@ def copy_os(args, template, target_dir):
     if args.fast_install:
         run_command("rm -rf {0}".format(args.statedir))
 
+def get_cmd_env(template):
```
Ah this one slipped by me but it this should have a test function too.

gtkramer replied:
Ahh, even this?  Ok  :)

bryteise commented on ister_test.py:
```diff
@@ -2023,7 +2089,7 @@ def get_template_good():
         template = ister.get_template("file://{0}/test.json"
                                       .format(os.getcwd()))
-        if template != {"test": 1}:
+        if template != {"test": 1, "SoftwareManager": "swupd"}:
```
Mind adding another test here for dnf?

gtkramer replied:
This tests whether when missing a SoftwareManager key that it gets added and defaults to swupd.  Not sure if it makes sense to have a dnf test for this?

bryteise replied:
Ah sorry, I was thinking of tests for validating that dnf and swupd are okay to use.

gtkramer replied:
I think I might have that covered with copy_os_switch_xxx (dnf and swupd)?

bryteise replied:
Ah yes indeed.

bryteise commented:
Almost there just a couple minor tweaks and I think we can merge it.

gtkramer commented:
I accidentally squashed the commit that did prep work for enabling dnf X( during a rebase.  I decided to squash one other commit to group work at a little bit of a higher level.  Should I re-separate anything?

I went through and tidied up a couple of little things for consistency with other tests and their description terminology.  I just ran through another round of testing and things still look good after the updates.

gtkramer commented:
@mdhorn are you doing a rewrite of the installer?  Where should I go to work on adding dnf support to the new installer?
